[train][docs] Add Azure Files to persistent storage docs (#63406)
ray-project/ray: doc/source/train/user-guides/persistent-storage.rst

Reviewer: ```suggestion
Use cloud object storage by specifying a storage container URI as the :class:`RunConfig(storage_path) <ray.train.RunConfig>`:
```

Azure doesn't use "bucket". Term storage container is more vendor neutral.

Author: Applied in 79f41e0. The intro line now reads "Use cloud object storage by specifying a storage container URI as ...".

Reviewer: ```suggestion
Cloud object storage
---------------------

The Ray team recommends using cloud object storage such as S3, GCS, or Azure Blob Storage to persist Ray Train checkpoint files.
```

Get rid of parentheses in titles

Author: Applied in 79f41e0 — heading is now "Cloud object storage" with no parenthesized provider list, and the recommendation moves into a plain sentence below the heading.

**Ray Train expects all workers to be able to write files to the same persistent storage location.**
Therefore, Ray Train requires some form of external persistent storage such as

> cloud storage (e.g., S3, GCS, Azure Blob Storage) or a shared filesystem

Author: Good catch — applied in 9785d06. The Cloud storage section header now reads "AWS S3, Google Cloud Storage, Azure Blob Storage" to match the summary, and the example block lists commented-out GCS and Azure Blob `storage_path` lines (mirroring how the Shared filesystem block lists HDFS and Azure Files). The Azure Blob example points at the existing "Custom storage" subsection since it needs a `storage_filesystem` (`adlfs`/`pyarrow.fs.PyFileSystem`) rather than a bare URI.

> cloud storage (e.g., S3, GCS, Azure Blob Storage) or a shared filesystem

Author: Updated in 36e72f7 — the Azure Blob example now uses the canonical `abfs://container@account.dfs.core.windows.net/sub-path/` URI (the scheme adlfs/fsspec and Hadoop's ABFS driver expose), matching the URI-style `s3://` and `gs://` examples above. Still gated behind `storage_filesystem` via `adlfs`/`PyFileSystem` per the Custom storage section.

> cloud storage (e.g., S3, GCS, Azure Blob Storage) or a shared filesystem

Reviewer: Noting that `abfss://` is the preferred driver for Azure, but I haven't checked Ray source code to confirm it's respected in all places.

> cloud storage (e.g., S3, GCS, Azure Blob Storage) or a shared filesystem

Reviewer: ```suggestion
cloud object storage (for example, S3, GCS, or Azure Blob Storage) or a shared filesystem
```

Author: Applied in 79f41e0.

> cloud storage (e.g., S3, GCS, Azure Blob Storage) or a shared filesystem

Author: Done in 79f41e0 — the Azure tab now uses `abfss://container@account.dfs.core.windows.net/sub-path/` (TLS-enforced) for the URI. Thanks for the pointer.

> (e.g., AWS EFS, Google Filestore, Azure Files, HDFS) for multi-node training.

Reviewer: ```suggestion
(for example, AWS EFS, Google Cloud Filestore, Azure Files, or HDFS) for multi-node training.
```

Small grammar/style edits.

Author: Applied in 79f41e0.

        ...,
        run_config=train.RunConfig(
            storage_path="s3://bucket-name/sub-path/",
            # GCS example:
            # storage_path="gs://bucket-name/sub-path/",
            # Azure Blob Storage example (requires a custom
            # ``storage_filesystem`` via ``adlfs`` /
            # ``pyarrow.fs.PyFileSystem``, see "Custom storage" below):

>             # storage_path="abfs://container@account.dfs.core.windows.net/sub-path/",

Reviewer: Suggestion: instead of commented-out URIs inside one example, use a tab-set so each cloud gets its own runnable example. This matches the per-cloud pattern already used in doc/source/data/loading-data.rst:196 and doc/source/data/saving-data.rst:90.
Sketch:

```
.. tab-set::

    .. tab-item:: AWS

        .. testcode::

            from ray import train

            trainer = train.Trainer(
                ...,
                run_config=train.RunConfig(
                    storage_path="s3://bucket-name/sub-path/",
                    name="experiment_name",
                )
            )

    .. tab-item:: Google Cloud

        .. testcode::
            :skipif: True

            from ray import train

            trainer = train.Trainer(
                ...,
                run_config=train.RunConfig(
                    storage_path="gs://bucket-name/sub-path/",
                    name="experiment_name",
                )
            )

    .. tab-item:: Azure

        Ray Train uses ``pyarrow.fs`` for storage I/O. Wrap ``adlfs.AzureBlobFileSystem``
        in a ``pyarrow.fs.PyFileSystem`` and pass it as ``storage_filesystem``:

        .. testcode::
            :skipif: True

            import adlfs
            from pyarrow.fs import PyFileSystem, FSSpecHandler
            from ray import train

            azure_fs = PyFileSystem(FSSpecHandler(
                adlfs.AzureBlobFileSystem(account_name="account-name")
            ))

            trainer = train.Trainer(
                ...,
                run_config=train.RunConfig(
                    storage_filesystem=azure_fs,
                    storage_path="abfss://container@account.dfs.core.windows.net/sub-path/",
                    name="experiment_name",
                )
            )

```

Author: Applied in 79f41e0 — the cloud section now uses a `.. tab-set::` with one tab per provider (AWS S3 / Google Cloud Storage / Azure Blob Storage), matching the pattern at `doc/source/data/loading-data.rst:196`. The Azure tab shows the full `adlfs.AzureBlobFileSystem` + `PyFileSystem(FSSpecHandler(...))` plumbing with an `abfss://` URI, and links to the Custom storage section for more on `storage_filesystem`. The shared filesystem section got the same treatment with Mounted / HDFS tabs.

            storage_path="/mnt/cluster_storage",
            # HDFS example:
            # storage_path=f"hdfs://{hostname}:{port}/subpath",
            # Azure Files example (mount the file share on every node first):
            # storage_path="/mnt/azure-fileshare",
            name="experiment_name",
        )
    )

Ensure that all nodes in the Ray cluster have access to the shared filesystem, e.g. AWS EFS, Google Cloud Filestore, Azure Files, or HDFS,
so that outputs can be saved to there.
In this example, all files are saved to ``/mnt/cluster_storage/experiment_name`` for further processing.

> 

Reviewer: ```suggestion
Shared filesystem
------------------

You can use shared filesystems such as HDFS or NFS, including AWS EFS, Azure Files, or Google Cloud Filestore. You can either mount these filesystems to make them available to nodes in your Ray cluster or use fully qualified URIs. In either case, you must ensure that networking rules and security permissions allow access from all nodes in your cluster.

Use by specifying the shared storage path as the :class:`RunConfig(storage_path) <ray.train.RunConfig>`:

.. tab-set::

    .. tab-item:: Mount

        .. testcode::
            :skipif: True

            from ray import train
            from ray.train.torch import TorchTrainer

            trainer = TorchTrainer(
                ...,
                run_config=train.RunConfig(
                    storage_path="/mnt/cluster_storage",
                    # HDFS example:
                    # storage_path=f"hdfs://{hostname}:{port}/subpath",
                    # Azure Files example (mount the file share on every node first):
                    # storage_path="/mnt/azure-fileshare",
                    name="experiment_name",
                )
            )

.. tab-set::

    .. tab-item:: HDFS

        .. testcode::
            :skipif: True

            from ray import train
            from ray.train.torch import TorchTrainer

            trainer = TorchTrainer(
                ...,
                run_config=train.RunConfig(
                    storage_path=f"hdfs://{hostname}:{port}/subpath",
                    name="experiment_name",
                )
            )

```

Author: Applied in 79f41e0 — the shared filesystem heading is now just "Shared filesystem", with the providers (AWS EFS, Google Cloud Filestore, Azure Files, HDFS, NFS) moved into the prose along with the mount-vs-URI and networking-permissions notes. The example is a `.. tab-set::` with "Mounted filesystem" and "HDFS" tabs (the Mounted tab covers AWS EFS / Filestore / Azure Files / NFS uniformly via the mount path). Collapsed your two adjacent `.. tab-set::` blocks into a single one — let me know if you'd prefer them split.
